File: src/main/scala/org/bakkes/game/controller/init/scripting/dsl/PokemonSpeciesDsl.java
package org.bakkes.game.controller.init.scripting.dsl;

import com.google.inject.Inject;
import com.google.inject.Provider;
import com.google.inject.name.Named;
import groovy.lang.Closure;
import org.bakkes.game.controller.init.scripting.dsl.anotation.Required;
import org.bakkes.game.controller.init.scripting.dsl.anotation.Result;
import org.bakkes.game.controller.init.scripting.loader.ScriptLoader;
import org.bakkes.game.model.pokemon.PokemonSpecies;
import org.bakkes.game.model.pokemon.PokemonStatistics;
import org.bakkes.game.model.pokemon.Type;
import org.newdawn.slick.util.Log;

import java.nio.file.Path;
import java.util.Collection;

public class PokemonSpeciesDsl extends ASpriteNamedDsl {

	private PokemonSpecies target = new PokemonSpecies();
	@Inject Provider<PokemonStatistics> statisticsProvider;
	@Inject ScriptLoader loader;
	@Inject @Named("moves") Path path;
	public final void setEvolution(final String evolution) {
		target.setEvolution(evolution);
	}
	public final void setName(final String name) {
		target.setName(name);
	}
	public final void setEvolutionLevel(final int evolutionLevel) {
		target.setEvolutionLevel(evolutionLevel);
	}
	public void base(final Closure<Void> commands){
		target.setBase(createFromCommands(commands));
	}
	public void onLevelIncrease(final Closure<Void> commands){
		target.setIncrease(createFromCommands(commands));
	}
	private PokemonStatistics createFromCommands(final Closure<Void> commands){
		final PokemonStatistics stats = statisticsProvider.get();
		delegate(commands, stats);
		return stats;
	}
	@Required
	public final void setType(final String type) {
		if(type.equals(Type.unkown.name())){
			Log.warn("the unkown type is only meant for errorhandeling, using it in production code may lead to problems");
		}
		for(final Type t : Type.values()){
			if(t.name().equals(type)){
                target.setType(t);
                return;
			}
		}
		Log.warn("type " + type + " not found, setting to unkown");
		target.setType(Type.unkown);
	}
	public final void setMoves(final Collection<String> moves) {
		for(final String moveName : moves){
			Log.info("reading " + moveName);
			final MoveDsl factory = new MoveDsl(moveName);
			loader.load(path.resolve(moveName + ".dsl"), factory);
			target.getMoves().add(factory.createMove());
		}
	}
	public void setTrainingSpeed(final float trainingSpeed) {
		target.setTrainingSpeed(trainingSpeed);
	}
	@Override
	@Result
	public PokemonSpecies getTarget(){
		return target;
	}

}

File: src/main/scala/org/bakkes/game/controller/init/scripting/dsl/PokemonSpeciesDsl.scala
package org.bakkes.game.controller.init.scripting.dsl

import com.google.inject.Inject
import com.google.inject.Provider
import com.google.inject.name.Named
import groovy.lang.Closure
import org.bakkes.game.controller.init.scripting.dsl.anotation.Required
import org.bakkes.game.controller.init.scripting.dsl.anotation.Result
import org.bakkes.game.controller.init.scripting.loader.{SpeciesLoader, ScriptLoader}
import org.bakkes.game.model.pokemon.PokemonSpecies
import org.bakkes.game.model.pokemon.PokemonStatistics
import org.bakkes.game.model.pokemon.Type
import org.newdawn.slick.util.Log
import java.nio.file.Path
import java.util.Collection

class PokemonSpeciesDsl @Inject()(
	private var target: PokemonSpecies,
	private val statisticsProvider: Provider[PokemonStatistics],
	private val scriptLoader: ScriptLoader,
	@Named("moves")
	private val path: Path,
	private val speciesLoader: SpeciesLoader
) extends ADsl{
	final def setEvolution(evolution: String) {
		target = target.copy(
			evolution = Some(speciesLoader.load(evolution))
		)
	}

	final def setName(name: String) {
		Log.warn("this name will be overriden: " + name + " to avoid confusion")
		target = target.copy(
			name = name
		)
	}

	final def setEvolutionLevel(evolutionLevel: Int) {
		target = target.copy(
			evolutionLevel = evolutionLevel
		)
	}

	def base(commands: Closure[Void]) {
		target = target.copy(
			base = createFromCommands(commands)
		)
	}

	def onLevelIncrease(commands: Closure[Void]) {
		target = target.copy(
			increase = createFromCommands(commands)
		)
	}

	private def createFromCommands(commands: Closure[Void]): PokemonStatistics = {
		val stats: PokemonStatistics = statisticsProvider.get
		delegate(commands, stats)
		return stats
	}

	@Required final def setType(element: String) {
		if (element == Type.unkown.name) {
			Log.warn("the unkown type is only meant for errorhandeling, using it in production code may lead to problems")
		}
		for (t <- Type.values) {
			if (t.name == element) {
				target = target.copy(
					element = t
				)
				return
			}
		}
		Log.warn("type " + element + " not found, setting to unkown")
		target = target.copy(
			element = Type.unkown
		)
	}

	final def setMoves(moves: Collection[String]) {
		import scala.collection.JavaConversions._
		for (moveName <- moves) {
			Log.info("reading " + moveName)
			val factory: MoveDsl = new MoveDsl(moveName)
			scriptLoader.load(path.resolve(moveName + ".dsl"), factory)
			target.copy(
				moves =  target.moves :+ factory.createMove
			)
		}
	}

	def setTrainingSpeed(trainingSpeed: Float) {
		target = target.copy(
			trainingSpeed = trainingSpeed
		)
	}

	@Result
	def getTarget: PokemonSpecies = target
}